src/Grpc.AspNetCore.Server/Internal/DefaultGrpcServiceActivator.cs

```diff
@@ -0,0 +1,55 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Diagnostics;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Grpc.AspNetCore.Server.Internal
+{
+    internal class DefaultGrpcServiceActivator<TGrpcService> : IGrpcServiceActivator<TGrpcService> where TGrpcService : class
+    {
+        private static readonly Lazy<ObjectFactory> _objectFactory = new Lazy<ObjectFactory>(() => ActivatorUtilities.CreateFactory(typeof(TGrpcService), Type.EmptyTypes));
+        private readonly IServiceProvider _serviceProvider;
+        private bool? _created;
```

Comment:
Dunno if this code is nearly complete, but the three-state boolean is confusing. To me, a single variable `_createPreviouslyCalled` or such looks much more readable. As is, only its property of having the value is used, not the value itself.

Reply:
The code is taken from https://github.com/aspnet/AspNetCore/blob/master/src/SignalR/server/Core/src/Internal/DefaultHubActivator.cs. The value is used in the dispose section for cleanup purposes. This brings up a good point on how we should handle TService that implement IDisposable for example.